### DISCUSSION TITLE
VTK or PyVista?

### DISCUSSION THREAD

**Discussion by gsal** [writer/writer-framework#74] [General]
Hello.

Hopefully only the first of many posts to come....I know it is a young project, but I am impressed; as dangerious as it may be, the graphical GUI builder makes this framework rather approachable by my not-programmer colleagues....if anything, they can at least mock what they want instead of hand-waving. 

Anyway, is it possible to do 3D rendering with VTK or PyVista?

**Comment by ramedina86** [comment]
Love this, quite on point with the essence of the project.

Regarding VTK/PyVista... I could try to wrap VTK, but I'm not familiar with it, so I'd likely end up producing something underwhelming.

For a Python web framework with VTK rendering I think your best bet is Trame, made by the company that makes VTK. It's very different from Streamsync and it doesn't have a UI editor, but seems to be tightly integrated with VTK.

**Reply by gsal** [reply]
Yes, thanks. 

I am new to the web development scene and not being a full-stack developer, I am evaluating these pure-python frameworks. Just by reading the docs, looking at how pythonic the code looks, how state management and callbacks are done...I am liking Streamsync and NiceGUI. 

Regarding support for VTK, I learned about Trame and Dash earlier this week; but neither is as approachable as Streamsync or NiceGUI, nor their VTK interface as approachable as PyVista.

Now, to be honest, I don't think I can handle a one-to-one interface to VTK; so, I much rather use PyVista.

From _your_ point of view, I would like to think that supporting PyVista shouldn't be that difficult, though I don't know what I am talking about. But we are not talking about implementing a brand new Python interface to some C++ library, PyVista is already Python; also, the docs show PyVista being used in JupyterLab, which is web-based.

So, maybe, just maybe, it's a matter of having some Streamsync widget that can handle/accept whatever pyvista.Plotter.mesh() or pyvista.Plotter.show() return?

Anyway, I sure hope you find it in your heart to support this ;-)

**Reply by ramedina86** [reply]
It's really encouraging to hear how much you want to use Streamsync, means a lot, especially considering how it's still on version 0.1.x.

Unfortunately it's not straightforward to embed PyVista. The interactive output of PyVista isn't HTML, it's an ipywidget, which requires a Jupyter Server (or Voila) to render it and manage interactions. There's no simple way to wrap this; if there was I'd be happy to.

You may have some luck with Solara, which supports ipywidgets. Again, it's not Streamsync and it doesn't have its visual editor, but it may help achieve what you're after.

If you _really_ want to use Streamsync (or NiceGUI), and your PyVista is just a component of many in your app, perhaps you can run it in a notebook using Voila and use an iframe to show it.

**Reply by gsal** [reply]
Fine; but don't believe for a second that this is going to stop me from using Streamsync for my non-3D project.  :-D

I will look into Solara, then; actually, I was able to do _something_ with Panel but ended up buggy.

'til the next one. 

Cheers.
